shell/src/main/java/org/apache/zeppelin/shell/ShellInterpreter.java

```diff
@@ -61,31 +66,62 @@ public void close() {}
   @Override
   public InterpreterResult interpret(String cmd, InterpreterContext contextInterpreter) {
     logger.debug("Run shell command '" + cmd + "'");
-    long start = System.currentTimeMillis();
     CommandLine cmdLine = CommandLine.parse("bash");
     cmdLine.addArgument("-c", false);
     cmdLine.addArgument(cmd, false);
     DefaultExecutor executor = new DefaultExecutor();
     ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
-    executor.setStreamHandler(new PumpStreamHandler(outputStream));
-
+    ByteArrayOutputStream errorStream = new ByteArrayOutputStream();
+    executor.setStreamHandler(new PumpStreamHandler(outputStream, errorStream));
     executor.setWatchdog(new ExecuteWatchdog(commandTimeOut));
+
+    Job runningJob = getRunningJob(contextInterpreter.getParagraphId());
+    Map<String, Object> info = runningJob.info();
+    info.put(EXECUTOR_KEY, executor);
     try {
-      int exitValue = executor.execute(cmdLine);
+      executor.execute(cmdLine);
       return new InterpreterResult(InterpreterResult.Code.SUCCESS, outputStream.toString());
     } catch (ExecuteException e) {
+      int exitValue = e.getExitValue();
       logger.error("Can not run " + cmd, e);
-      return new InterpreterResult(Code.ERROR, e.getMessage());
+      Code code = Code.ERROR;
+      String msg = errorStream.toString();
+      if (exitValue == 143) {
+        code = Code.INCOMPLETE;
+        msg = msg + "Paragraph received a SIGTERM";
+      }
+      return new InterpreterResult(code, msg);
     } catch (IOException e) {
       logger.error("Can not run " + cmd, e);
       return new InterpreterResult(Code.ERROR, e.getMessage());
     }
   }
 
-  @Override
-  public void cancel(InterpreterContext context) {}
+  private Job getRunningJob(String paragraphId) {
+    Job foundJob = null;
+    Collection<Job> jobsRunning = getScheduler().getJobsRunning();
+    for (Job job : jobsRunning) {
+      if (job.getId().equals(paragraphId)) {
+        foundJob = job;
+      }
+    }
+    return foundJob;
+  }
 
   @Override
+  public void cancel(InterpreterContext context) {
+    Job runningJob = getRunningJob(context.getParagraphId());
+    if (runningJob != null) {
+      Map<String, Object> info = runningJob.info();
+      Object object = info.get(EXECUTOR_KEY);
+      if (object != null) {
+        Executor executor = (Executor) object;
+        ExecuteWatchdog watchdog = executor.getWatchdog();
+        watchdog.destroyProcess();
```

Comment:
The DefaultExecutor#execute waits on the thread created by watchdog. Based on the exitstatus of the terminated process, the executor throws the exception.
Please let me know of any other way to terminate the running job.